### PR TITLE
Return an instance of a public row classes from certain API calls

### LINKED ISSUE #8668: Expose a new row-like object API for the data grids
Currently, grid APIs which return "row objects" are returning the actual row component from current state of the DOM.
However this is far from useful as with the current implementation virtualization state is not taken into account (#6158) as well 
as exposing API which should not be usable by users of the grid.

As a solution, a new row-like interface should be exposed by the grids as the type returned from certain API calls.
This new interface will act as a façade for the core row API while also taking the current virtualization state into account.

### PR DESCRIPTION
**Pull request by hanastasov**
Closes #8668 

### Additional information (check all that apply):
 - [x] Bug fix
 - [x] New functionality
 - [x] Documentation
 - [x] Demos
 - [ ] CI/CD

### Checklist:
 - [x] All relevant tags have been applied to this PR
 - [x] This PR includes unit tests covering all the new code ([test guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Test-implementation-guidelines-for-Ignite-UI-for-Angular))
 - [x] This PR includes API docs for newly added methods/properties ([api docs guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Documentation-Guidelines))
 - [ ] This PR includes `feature/README.MD` updates for the feature docs
 - [ ] This PR includes general feature table updates in the root `README.MD`
 - [x] This PR includes `CHANGELOG.MD` updates for newly added functionality
 - [x] This PR contains breaking changes
 - [x] This PR includes `ng update` migrations for the breaking changes ([migrations guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Update-Migrations))
 - [x] This PR includes behavioral changes and the feature specification has been updated with them

### CODE REVIEW

File: projects/igniteui-angular/src/lib/grids/grid-base.directive.ts [hunk 1]
```diff
@@ -5279,6 +5260,19 @@ export abstract class IgxGridBaseDirective extends DisplayDensityBase implements
...
+     * @hidden @internal
+     */
+    public get allRowsData(): any[] {
+        return this._allRowsData;
+    }
+    /**
+     * @hidden @internal
+     */
+    public set allRowsData(data: any[]) {
+        this._allRowsData = data;
```
Comment (rkaraivanov) [inline projects/igniteui-angular/src/lib/grids/grid-base.directive.ts:5273]: Isn't these the same as `dataView`

File: projects/igniteui-angular/src/lib/grids/grid-public-row.ts [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+     * let selectedRowData = this.grid.selectedRows[0].rowData;
+     * ```
+     */
+    public get data(): any {
+        const grid = this.grid;
+        let allRowsData = grid.allRowsData;
+        if (!this._data) {
+            if (grid.pinnedRecordsCount) {
+                allRowsData = grid.pinning.rows !== RowPinningPosition.Bottom ?
+                    [...grid.pinnedRecords, ...grid.allRowsData] : [...grid.allRowsData, ...grid.pinnedRecords];
```
Comment (rkaraivanov) [inline projects/igniteui-angular/src/lib/grids/grid-public-row.ts:35]: Maybe `grid.dataView[this.index]`?

File: projects/igniteui-angular/src/lib/grids/grid-public-row.ts [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+            }
+        }
+        return this._data ?? allRowsData[this.index];
+    }
+
+    /**
+     * Needs to be removed !
```
Comment (rkaraivanov) [inline projects/igniteui-angular/src/lib/grids/grid-public-row.ts:42]: 👍

File: projects/igniteui-angular/src/lib/grids/grid-public-row.ts [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+        this.grid.cdr.markForCheck();
+    }
+
+    /**
+     * Sets whether this specific row has disabled functionality for editing and row selection.
+     * Default value is `false`.
+     * ```typescript
+     * const isDisabled = row.disabled;
+     * ```
+     */
```
Comment (rkaraivanov) [inline projects/igniteui-angular/src/lib/grids/grid-public-row.ts:95]: I guess the disabled state is active when a row is pinned and users have the 'ghost' instance from the unpinned area, right?
If that is the case, then this doc string should be more specific when and why a row is in a disabled state.

File: projects/igniteui-angular/src/lib/grids/grid-public-row.ts [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Returns the child rows. Always returns null for IgxGridRow.
+     */
+    public get children(): any[] {
+        return null;
```
Comment (rkaraivanov) [inline projects/igniteui-angular/src/lib/grids/grid-public-row.ts:129]: Typing can be more specific here, e.g. 
```typescript
get children(): null | RowType[] // <-- or even more specific 
```

File: projects/igniteui-angular/src/lib/grids/grid-public-row.ts [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+    public set expanded(val: boolean) {
+        this.gridAPI.set_row_expansion_state(this.rowID, val);
+    }
+
+    /**
+     * Get a reference to the grid that contains the selected row.
+     */
+    protected get grid(): IgxGridComponent | IgxTreeGridComponent | IgxHierarchicalGridComponent {
+        return this._grid as IgxGridComponent;
+    }
```
Comment (rkaraivanov) [inline projects/igniteui-angular/src/lib/grids/grid-public-row.ts:155]: This seems a tad unnecessary.
You can achieve the same directly in the constructor
```typescript
constructor(protected grid: <TypeInfo>, ...) {}
```

File: projects/igniteui-angular/src/lib/grids/grid-public-row.ts [hunk 7]
```diff
@@ -0,0 +1,492 @@
...
+            }
+        }
+        return this._data ?? allRowsData[this.index];
+    }
+
+    /**
+     * Returns the view index calculated per the grid page.
+     */
+    public get viewIndex(): number {
+        if ((this.grid as any).groupingExpressions.length) {
```
Comment (rkaraivanov) [inline projects/igniteui-angular/src/lib/grids/grid-public-row.ts:278]: Does the Tree grid have `groupingExpressions`?
Please, clarify this check.

File: projects/igniteui-angular/src/lib/grids/grid-public-row.ts [hunk 8]
```diff
@@ -0,0 +1,492 @@
...
+    /**
+     * The `ITreeGridRecord` passed to the row component.
+     *
+     * ```typescript
+     * const row = this.treeGrid.getRowByKey(1) as IgxTreeGridRowComponent;
+     * const treeRow = row.treeRow;
+     * ```
+     */
+    public get treeRow(): ITreeGridRecord {
+        const treeRow = this._treeRow ?? this._tgrid.records.get(this.rowID);
```
Comment (rkaraivanov) [inline projects/igniteui-angular/src/lib/grids/grid-public-row.ts:332]: You may directly return here

**Review by rkaraivanov — DISMISSED** [review]

File: projects/igniteui-angular/src/lib/grids/common/events.ts [hunk 9]
```diff
@@ -170,11 +171,6 @@ export interface IRowToggleEventArgs extends IBaseEventArgs {
  * Event emitted when a row's pin state changes.
  */
 export interface IPinRowEventArgs extends IBaseEventArgs {
-    /**
-     * The row component instance, that was pinned/unpinned.
-     * May be undefined if row does not exist in the current visible data.
-     */
-    readonly row?: IgxRowDirective<IgxGridBaseDirective & GridType>;
```
[the remark this answers is not in the document]
Comment (hanastasov) [reply projects/igniteui-angular/src/lib/grids/common/events.ts:177]: Will do it.

File: projects/igniteui-angular/src/lib/grids/grid-public-row.ts [hunk 10]
```diff
@@ -28,15 +27,7 @@ export class IgxGridRow implements RowType {
...
-        const grid = this.grid;
-        let allRowsData = grid.allRowsData;
-        if (!this._data) {
-            if (grid.pinnedRecordsCount) {
-                allRowsData = grid.pinning.rows !== RowPinningPosition.Bottom ?
-                    [...grid.pinnedRecords, ...grid.allRowsData] : [...grid.allRowsData, ...grid.pinnedRecords];
-            }
-        }
-        return this._data ?? allRowsData[this.index];
+        return this._data ?? this.grid.data[this.index];
```
Comment (rkaraivanov) [inline projects/igniteui-angular/src/lib/grids/grid-public-row.ts:30]: I guess this is a typo, but this should be `this.grid.dataView[this.index]` right?

**Review by zdrawku — APPROVED** [review]

**Pull request merged**